numpy/core/src/umath/string_ufuncs.cpp

```diff
@@ -0,0 +1,448 @@
+         * does not support unicode whitespace (and never has).
+         */
+        while (len1 > 0) {
+            character c = str1[len1-1];
+            if (c != (character)0 && !NumPyOS_ascii_isspace(c)) {
+                break;
+            }
+            len1--;
+        }
+        while (len2 > 0) {
+            character c = str2[len2-1];
+            if (c != (character)0 && !NumPyOS_ascii_isspace(c)) {
+                break;
+            }
+            len2--;
+        }
+    }
+
+    int n = PyArray_MIN(len1, len2);
+
+    for (int i = 0; i < n; i++) {
+        int cmp = character_cmp(*str1, *str2);
+        if (cmp != 0) {
+            return cmp;
+        }
+        str1++;
+        str2++;
+    }
+    if (len1 > len2) {
+        for (int i = n; i < len1; i++) {
+            int cmp = character_cmp(*str1, (character)0);
+            if (cmp != 0) {
+                return cmp;
+            }
+            str1++;
+        }
+    }
+    else if (len2 > len1) {
+        for (int i = n; i < len2; i++) {
+            int cmp = character_cmp((character)0, *str2);
+            if (cmp != 0) {
+                return cmp;
+            }
+            str2++;
+        }
+    }
+    return 0;
+}
+
+
+template <bool rstrip, int comp, typename character>
+static int
+string_comparison_loop(PyArrayMethod_Context *context,
+        char *const data[], npy_intp const dimensions[],
+        npy_intp const strides[], NpyAuxData *NPY_UNUSED(auxdata))
+{
+    /*
+     * Note, this works in CPython even without the GIL, however it may be that
+     * this will have to be moved into `auxdata` eventually, which may be
+     * slightly faster/cleaner (but also slightly more involved) in any case.
+     */
+    int len1 = context->descriptors[0]->elsize / sizeof(character);
+    int len2 = context->descriptors[1]->elsize / sizeof(character);
+
+    char *in1 = data[0];
+    char *in2 = data[1];
+    char *out = data[2];
+
+    npy_intp N = dimensions[0];
+
+    while (N--) {
+        int cmp = string_cmp<rstrip>(
+                len1, (character *)in1, len2, (character *)in2);
+        npy_bool res;
+        if (comp == Py_EQ) {
+            res = cmp == 0;
+        }
+        else if (comp == Py_NE) {
+            res = cmp != 0;
+        }
+        else if (comp == Py_LT) {
+            res = cmp < 0;
+        }
+        else if (comp == Py_LE) {
+            res = cmp <= 0;
+        }
+        else if (comp == Py_GT) {
+            res = cmp > 0;
+        }
+        else if (comp == Py_GE) {
+            res = cmp >= 0;
+        }
+        else {
+            assert(0);
+        }
+        *(npy_bool *)out = res;
+
+        in1 += strides[0];
+        in2 += strides[1];
+        out += strides[2];
+    }
+    return 0;
+}
+
+
+/*
+ * Machinery to add the string loops to the existing ufuncs.
+ */
+
+/*
+ * This function replaces the strided loop with the passed in one,
+ * and registers it with the given ufunc.
+ */
+static int
+add_loop(
```

Comment:
Total nitpick, but why the new line?

Reply:
ah, because right now I tend to 8-space hanging indent, but that feels strange if there is more than one "special" parameter on the other line (and it is 1 space off here).  Anyway, just removed it, and aligned the second line with the bracket here...